homeassistant/components/fritzbox/binary_sensor.py

```diff
@@ -68,18 +68,26 @@ async def async_setup_entry(
     hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
 ) -> None:
     """Set up the FRITZ!SmartHome binary sensor from ConfigEntry."""
-    coordinator: FritzboxDataUpdateCoordinator = hass.data[FRITZBOX_DOMAIN][
-        entry.entry_id
-    ][CONF_COORDINATOR]
-
-    async_add_entities(
-        [
-            FritzboxBinarySensor(coordinator, ain, description)
-            for ain, device in coordinator.data.devices.items()
-            for description in BINARY_SENSOR_TYPES
-            if description.suitable(device)
-        ]
-    )
+    coordinator: FritzboxDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        CONF_COORDINATOR
+    ]
+    added_devices: list[str] = []
+
+    def _add_entities() -> None:
+        """Add devices."""
+        entities: list[FritzboxBinarySensor] = []
+        for ain, device in coordinator.data.devices.items():
+            if ain in added_devices:
+                continue
+            added_devices.append(ain)
+            for description in BINARY_SENSOR_TYPES:
+                if description.suitable(device):
+                    entities.append(FritzboxBinarySensor(coordinator, ain, description))
+        async_add_entities(entities)
+
+    entry.async_on_unload(coordinator.async_add_listener(_add_entities))
```

Comment:
I missed this. You will trigger this on each data update. It will iterate ALL devices in each platform on every data update, this is too much.

Your old new_devices solution was actually better. Sorry :)

The alternative would be to have _add_devices take a parameter with a list of ains to add. Then have coordinator keep track of when a new device shows up and trigger a dispatcher event for this. (I thought you already used a dispatcher event).

But.. since you are piggy backing on the coordinator listener your old solution could have worked. We generally dont use that for non entity updates, but i dont see why we cant. So might be better to revert to your old version.

Reply:
what do you think about this solution: e511040 (_only applied for binary sensors as showcase, when agreed i'll add it to all platforms_)